packages/compiler/test/checker/deprecation.test.ts

```diff
@@ -0,0 +1,199 @@
+import { Diagnostic } from "../../src/index.js";
+import {
+  BasicTestRunner,
+  createTestRunner,
+  expectDiagnosticEmpty,
+  expectDiagnostics,
+} from "../../src/testing/index.js";
+
+describe("compiler: checker: deprecation", () => {
+  let runner: BasicTestRunner;
+
+  beforeEach(async () => {
+    runner = await createTestRunner();
+  });
+
+  function expectDeprecations(diagnostics: readonly Diagnostic[], deprecations: string[]) {
+    expectDiagnostics(
+      diagnostics,
+      deprecations.map((x) => ({
+        code: "deprecated",
+        message: `Deprecated: ${x}`,
+        severity: "warning",
+      }))
+    );
+  }
+
+  describe("#deprecated directive", () => {
+    it("emits deprecation for use of deprecated model types", async () => {
+      const diagnostics = await runner.diagnose(`
+          #deprecated "OldFoo is deprecated"
+          model OldFoo {}
+
+          op get(): OldFoo;
+
+          model Bar {
+            foo: string | OldFoo;
+          }
+          `);
+
+      expectDeprecations(diagnostics, ["OldFoo is deprecated", "OldFoo is deprecated"]);
+    });
+
+    it("emits deprecation for use of model that extends deprecated parent", async () => {
+      const diagnostics = await runner.diagnose(`
+          #deprecated "OldFoo is deprecated"
+          model OldFoo {}
+
+          model NewFoo extends OldFoo {} // Diagnostic here
+          model IsFoo is NewFoo {} // Diagnostic here
+          model NewIsFoo is IsFoo {} // No diagnostic here
+          `);
+
+      expectDeprecations(diagnostics, ["OldFoo is deprecated", "OldFoo is deprecated"]);
+    });
+
+    it("emits deprecation for use of deprecated model properties", async () => {
+      const diagnostics = await runner.diagnose(`
+          model Foo {
+            #deprecated "Use id instead"
+            name: string;
+          }
+
+          model Bar {
+            value: Foo.name;
+          }
+
+          op get(name: Foo.name): string;
+          `);
+
+      expectDeprecations(diagnostics, ["Use id instead", "Use id instead"]);
+    });
+
+    it("emits deprecation for use of deprecated model properties from extended models", async () => {
+      const diagnostics = await runner.diagnose(`
+          model Foo {
+            #deprecated "Use id instead"
+            name: string;
+          }
+
+          model Bar extends Foo {}
+          model Baz is Foo {}
+          model Buz { ...Foo }
+
+          op get(name: Bar.name): string; // Diagnostic here
+          op put(name: Baz.name): string; // No diagnostic here
```

Comment:
shouldn't that be deprecated? same as the next one

Reply:
I didn't see any note about this in the proposal or implementation issue, so I wasn't sure.  My logic is that `is` and spread are basically copying the properties across so the same deprecation semantics might not apply in a newly constructed model.  If that's not the case, I'll be happy to change it.

Reply:
Turns out that if I move `checkDirectives` into `createType` then the deprecation directive gets carried across, making the copied properties deprecated as well.  If we want that, then I'll keep that behavior in.

Reply:
I think basically the behavior you want: 
if you reference a deprecated type it should complain and not carry it over(that include on the model for `model is`, interface for `interface extends`)

However the member of that types have not been referenced dirrectly so their deprecation should carry over.

See behavior in typescript 
https://www.typescriptlang.org/play?#code/PQKhAIAEBMFMAcBOsDGBDALra4TAFACWAdlogGZoqzgCC4A3uAL75GmwVU0BCj+4QeFAQYCZOiw48AoWgBc4AM4ZEJAOYBufKzYkylauADC4WAA8sxaEvB8GOtuwPdwAFQCMZy7Gu36TMzgzpyGNJ4A+gBM3lY27l6BwsDgABacNMQA9hjgcPAZkthsGACeBe4xALx2ANoARGj1ALraZRVuAMzgNcYNTa1sQA

Reply:
That example only accounts for `extends` though (and it makes sense in that case), my question is whether the deprecation of a property should be carried across in a case where the properties are being copied into a new type via `is` and spread.  What do you think should happen in that case/

Reply:
Thanks!